pkg/daemon/apply.go

```diff
@@ -0,0 +1,213 @@
+package daemon
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/golang/glog"
+	"github.com/coreos/go-systemd/dbus"
+	igntypes "github.com/coreos/ignition/config/v2_2/types"
+	mapset "github.com/deckarep/golang-set"
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	"k8s.io/kubectl/pkg/drain"
+)
+
+type actionResult interface {
+	Describe(dn *Daemon) string
+	Execute(dn *Daemon, newConfig *mcfgv1.MachineConfig) error
+}
+
+type NoOpPostAction struct {
+	actionResult
+}
+
+func (a NoOpPostAction)Describe(dn *Daemon) string {
+	return "No action needed to apply update"
+}
+
+func (a NoOpPostAction)Execute(dn *Daemon, newConfig *mcfgv1.MachineConfig) error {
+	return nil
+}
+
+type RebootPostAction struct {
+	actionResult
+
+	Reason string
+}
+
+func (a RebootPostAction)Describe(dn *Daemon) string {
+	return fmt.Sprintf("Rebooting node %v: %v", dn.node.GetName(), a.Reason)
+}
+
+func (a RebootPostAction)Execute(dn *Daemon, newConfig *mcfgv1.MachineConfig) error {
+	return dn.drainAndReboot(newConfig)
+}
+
+type DrainPostAction struct {
+	actionResult
+}
+
+func (a DrainPostAction)Describe(dn *Daemon) string {
+	return fmt.Sprintf("Draining node %v", dn.node.GetName())
+}
+
+func (a DrainPostAction)Execute(dn *Daemon, newConfig *mcfgv1.MachineConfig) error {
+	return dn.drain()
+}
+
+type ServicePostAction struct {
+	actionResult
+
+	Reason string
+
+	ServiceName string
+	ServiceAction string
+}
+
+
+func (a ServicePostAction)Describe(dn *Daemon) string {
+	return fmt.Sprintf("Restarting service %v", a.Reason)
+}
+
+func (a ServicePostAction)Execute(dn *Daemon, newConfig *mcfgv1.MachineConfig) error {
+	// TODO: add support for reload operation
+	// For now only restart operation is supported
+	systemdConnection, dbusConnErr := dbus.NewSystemConnection()
+	if dbusConnErr != nil {
+		glog.Warningf("Unable to establish systemd dbus connection: %s", dbusConnErr)
+		return dbusConnErr
+	}
+
+	defer systemdConnection.Close()
+
+	var err error
+	outputChannel := make(chan string)
+	switch a.ServiceAction {
+	case "restart":
+		glog.Infof("Restarting unit %q", a.ServiceName)
+		_, err = systemdConnection.RestartUnit(a.ServiceName, "replace", outputChannel)
+	case "stop":
+		glog.Infof("Stopping unit %q", a.ServiceName)
+		_, err = systemdConnection.StopUnit(a.ServiceName, "replace", outputChannel)
+	default:
+		return fmt.Errorf("Unhandled systemd action %q for %q", a.ServiceAction, a.ServiceName)
+	}
+
+	if err != nil {
+		return fmt.Errorf("Running systemd action failed: %s", err)
+	}
+
+	output := <-outputChannel
+	switch output {
+	case "done":
+		fallthrough
+	case "skipped":
+		glog.Infof("Systemd action successful: %s", output)
+	default:
+		return fmt.Errorf("Systemd action %s", output)
```

Comment:
This error message could use a little more context. Is the action unexpected? Is it invalid? What about this action is bad and/or erroneous?

Reply:
Messages have been improved and `skipped` is now treated as an error until there is reason not to